feat: offload cell proof computation from beacon node

hughy opened the pull request:
**Motivation**

Uses cell proofs from EL `getPayloadsV5` to offload cell proof construction

Refer to [EIP-7594 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7594.md#networking) for more on requiring blob transaction senders to compute cell proofs.

[Pending spec changes](https://github.com/ethereum/execution-apis/blob/cad4194e3fa37359a1be95e4aad2752d69691077/src/engine/osaka.md#engine_getpayloadv5) define updates for `getPayloadsV5` in the execution engine API

**Description**

* add engine_getPayloadV5 to engine API
* avoid computing cell proofs in computeDataColumnSidecars
* rename CELLS_PER_BLOB to CELLS_PER_EXT_BLOB to match spec
* update sszTypes for Fulu BlockContents, SignedBlockContents

Closes #7669

**Other notes**

* implement blobsBundle validation

I've added a validation function for validating `BlobsBundleV2` by computing cells and batch verifying the cell proofs, but don't currently call this function. We could validate this data on receiving responses from the EL or when producing the block body, but we might consider data from the EL trustworthy and skip costly verification.

hughy committed:
add engine_getPayloadV5 to engine API

- update getPayload with engine_getPayloadV5 on Fulu fork
- add validation to parseBlobsBundle
	- the structure of BlobsBundleV2 is identical to BlobsBundleV1 but
	  contains one proof per cell per blob instead of one per blob
	- validate lengths of blobs bundle properties based on fork

hughy committed:
avoid computing cell proofs in computeDataColumnSidecars

- use ckzg to compute cells from blob, use cell proofs from block contents
- cells proofs computed by tx sender, fetched from EL by engine_getPayloadsV5,
  and verified by CL
- update tests to use cell proofs as inputs to computeDataColumnSidecars

hughy committed:
move BlobsBundleV2 validation to separate module

separate module function validateBlobsAndCellProofs matches approach used for
BlobsBundleV1, which uses validateBlobsAndKzgCommitments module for validation

remove validation from parseBlobsBundle

hughy committed:
rename CELLS_PER_BLOB to CELLS_PER_EXT_BLOB to match spec

'EXT' clarifies that the value refers to the number of cells per 'extended'
blob

hughy committed:
validates blobsBundle during block production

hughy committed:
move blobsBundle validation to validation module

add unit tests for blobsBundle validation

hughy committed:
update sszTypes for Fulu BlockContents, SignedBlockContents

kzgProofs in block contents should contain one KZG proof per cell per blob

hughy committed:
minor branch cleanup

hughy committed:
add ckzg binding for compute_cells

g11tech commented on packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:
```diff
@@ -342,7 +343,18 @@ export async function produceBlockBody<T extends BlockType>(
           }
 
-          if (ForkSeq[fork] >= ForkSeq.deneb) {
+          if (ForkSeq[fork] >= ForkSeq.fulu) {
```
you don't need to add/dup  this code segment, fork >=deneb will already do the code flow, only now validateBlobsAndKzgCommitments needs to be fork aware if it batch verifies the proofs as well

hughy replied:
Sounds good, I'll update `validateBlobsAndKzgCommitments` and remove this segment

matthewkeil replied:
agreed!

g11tech commented on packages/beacon-node/src/chain/validation/blobsBundle.ts:
```diff
@@ -0,0 +1,41 @@
+ * Validate blobsBundle according to spec
+ * https://github.com/ethereum/execution-apis/blob/cad4194e3fa37359a1be95e4aad2752d69691077/src/engine/osaka.md#specification
+ */
+export function validateBlobsBundle(fork: ForkName, blobsBundle: BlobsBundle): void {
```
i think this can can be integrated to `validateBlobsAndKzgCommitments` by making it forkaware i.e. pass `fork: ForkBlobs` as the first arg to it and integrate the new logic there only in verification of the proofs because they are now cell proods

g11tech commented on packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:
```diff
@@ -342,7 +343,18 @@ export async function produceBlockBody<T extends BlockType>(
+
+            (blockBody as fulu.BeaconBlockBody).blobKzgCommitments = blobsBundle.commitments;
+            const blockHash = toRootHex(executionPayload.blockHash);
+            const contents = {kzgProofs: blobsBundle.proofs, blobs: blobsBundle.blobs};
```
for fulu's contents (so make contents fork aware type), also compute and directly add cells here, because cells will used else where and its better to just do the compute one time doesn't matter how easy it might be

also this will make things handy

g11tech commented on packages/beacon-node/src/chain/validation/blobsBundle.ts:
```diff
@@ -0,0 +1,41 @@
+    const cellIndices = Array.from({length: blobsBundle.blobs.length}).flatMap(() =>
+      Array.from({length: CELLS_PER_EXT_BLOB}, (_, i) => i)
+    );
+    const cells = blobsBundle.blobs.flatMap((blob) => ckzg.computeCells(blob));
```
contents should already have cells computed in the produceblockbody segment so that recomputes are not needed

g11tech commented on packages/beacon-node/src/chain/validation/blobsBundle.ts:
```diff
@@ -0,0 +1,41 @@
+    );
+  }
+
+  if (ForkSeq[fork] >= ForkSeq.fulu) {
```
i see this is already forkaware so better integrate it back to validateBlobsAndKzgCommitments or merge it here because both are doing the same thing

g11tech commented on packages/beacon-node/src/util/blobs.ts:
```diff
@@ -74,7 +74,7 @@ export function computeBlobSidecars(
   config: ChainForkConfig,
   signedBlock: SignedBeaconBlock,
-  contents: deneb.Contents & {kzgCommitmentsInclusionProof?: fulu.KzgCommitmentsInclusionProof}
+  contents: fulu.Contents & {kzgCommitmentsInclusionProof?: fulu.KzgCommitmentsInclusionProof}
```
make the contents forkaware and have the fulu ones store the cells computed once in produceblockbody...

hughy replied:
Is there any concern that adding the cells to block contents will differ from the [spec](https://github.com/ethereum/beacon-APIs/blob/2b1d7b5ac4756881bd29e7adacc9b7032343d981/types/fulu/block_contents.yaml#L9)?

matthewkeil replied:
I'm going to reach out to @nflaig and maybe we should update the api spec to have columns passed around instead of the blobs.  I am not a big fan of deviating from the spec containers for this so looking at updating the spec to avoid doing double calculation of the cells (before and after going to the validator for signature after production)

g11tech commented on packages/beacon-node/src/util/blobs.ts:
```diff
@@ -83,22 +83,26 @@ export function computeDataColumnSidecars(
-  const cellsAndProofs = blobs.map((blob) => ckzg.computeCellsAndKzgProofs(blob));
+  const {blobs, kzgProofs} = contents;
+  const cellsAndProofs = Array.from({length: blobs.length}, (_, rowNumber) => {
+    const cells = ckzg.computeCells(blobs[rowNumber]);
```
reuse cells

g11tech commented on packages/beacon-node/src/util/blobs.ts:
```diff
@@ -83,22 +83,26 @@ export function computeDataColumnSidecars(
-    const column = Array.from({length: blobs.length}, (_, rowNumber) => cellsAndProofs[rowNumber][0][columnIndex]);
-    const kzgProofs = Array.from({length: blobs.length}, (_, rowNumber) => cellsAndProofs[rowNumber][1][columnIndex]);
+    const column = Array.from({length: blobs.length}, (_, rowNumber) => cellsAndProofs[rowNumber].cells[columnIndex]);
+    const columnKzgProofs = Array.from({length: blobs.length}, (_, rowNumber) => cellsAndProofs[rowNumber].proofs[columnIndex]);
```
name can stay kzgProofs to keep things concise here and in the return

matthewkeil replied:
I've done enough double takes with the last couple PR's about column vs blob proofs... Would love to set a precedent in the codebase for being explicit about which are blob and which are column proofs so they dont get accidentally mixed up anywhere.  What do you think @g11tech ?

g11tech commented on packages/beacon-node/test/unit/db/api/repositories/dataColumn.test.ts:
```diff
@@ -50,14 +50,15 @@ describe("block archive repository", () => {
     const allDataColumnSidecars = computeDataColumnSidecars(config, singedBlock, {
       blobs: [blob, blob, blob],
-      kzgProofs: [commitment, commitment, commitment],
+      kzgProofs: Array.from({length: 3 * NUMBER_OF_COLUMNS}, () => kzgProof),
```
nice catch but sizes might be same i guess before for typescript to not flag it

hughy replied:
Now that `computeDataColumnSidecars` reuses the proofs from the contents instead of computing them the input proofs need to contain one per cell. This test fails if there is only one proof per blob

hughy committed:
move blobsBundle validation logic into validateBlobsAndKzgCommitments

make validateBlobsAndKzgCommitments fork aware

add fulu to ForkBlobs

always call validateBlobsAndKzgCommitments on fulu: client should always verify
proofs

hughy commented on packages/beacon-node/src/execution/engine/http.ts:
```diff
@@ -416,13 +416,15 @@ export class ExecutionEngineHttp implements IExecutionEngine {
-          : ForkSeq[fork] >= ForkSeq.capella
-            ? "engine_getPayloadV2"
-            : "engine_getPayloadV1";
+      ForkSeq[fork] >= ForkSeq.fulu
```
This nested conditional is getting quite nested, consider a switch

hughy committed:
rename columnKzgProofs to kzgProofs

hughy committed:
convert nested conditional in getPayload to switch statement

hughy committed:
fix denebSsz dependencies in fulu sszTypes

hughy committed:
fix lint errors in validateBlobsAndKzgCommitments

nflaig commented on packages/params/src/forkName.ts:
```diff
@@ -112,10 +112,10 @@ export function isForkPostFulu(fork: ForkName): fork is ForkPostFulu {
 export type ForkPostBlobs = ForkPostFulu;
-export type ForkBlobs = ForkName.deneb | ForkName.electra;
-export const forkBlobs = [ForkName.deneb, ForkName.electra];
+export type ForkBlobs = ForkName.deneb | ForkName.electra | ForkName.fulu;
```
why do we still have this, I thought we got rid of it in https://github.com/ChainSafe/lodestar/pull/7441 cc @matthewkeil

matthewkeil replied:
that was merging into unstable.  peerDAS Still has this as does my refactor.  There are tons of guards for this check but it should not have fulu.  just deneb and electra and should be an alias to ForkPreFulu and ForkPostDeneb

matthewkeil replied:
Agreed that we should not update this though.  We can use `ForkPostDeneb` for all the locations in this PR

hughy replied:
The `peerDAS` branch doesn't have that change yet, so this will need to be updated after `unstable` merges into `peerDAS`

matthewkeil commented on packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:
```diff
@@ -347,8 +348,8 @@ export async function produceBlockBody<T extends BlockType>(
-            if (this.opts.sanityCheckExecutionEngineBlobs) {
-              validateBlobsAndKzgCommitments(executionPayload, blobsBundle);
+            if (ForkSeq[fork] >= ForkSeq.fulu || this.opts.sanityCheckExecutionEngineBlobs) {
+              validateBlobsAndKzgCommitments(fork as ForkBlobs, executionPayload, blobsBundle);
```
```suggestion
              validateBlobsAndKzgCommitments(fork as ForkPostDeneb, executionPayload, blobsBundle);
```

matthewkeil commented on packages/beacon-node/src/chain/produceBlock/validateBlobsAndKzgCommitments.ts:
```diff
@@ -1,16 +1,49 @@
-export function validateBlobsAndKzgCommitments(_payload: ExecutionPayload, blobsBundle: BlobsBundle): void {
-  // sanity-check that the KZG commitments match the blobs (as produced by the execution engine)
+export function validateBlobsAndKzgCommitments(
+  fork: ForkBlobs,
```
```suggestion
  fork: ForkPostDeneb,
```

matthewkeil commented on packages/beacon-node/src/chain/produceBlock/validateBlobsAndKzgCommitments.ts:
```diff
@@ -1,16 +1,49 @@
+        `Invalid proofs length for BlobsBundleV1 format: expected ${blobsBundle.blobs.length}, got ${blobsBundle.proofs.length}`
+      );
+    }
+  }
```
Do you mind if we switch around this conditional and returning early to reduce the nesting depth?

```suggestion
  if (ForkSeq[fork] < ForkSeq.fulu) {
    if (blobsBundle.proofs.length !== blobsBundle.blobs.length) {
      throw new Error(
        `Invalid proofs length for BlobsBundleV1 format: expected ${blobsBundle.blobs.length}, got ${blobsBundle.proofs.length}`
      );
    }
    return;
  }

  const expectedProofsLength = blobsBundle.blobs.length * CELLS_PER_EXT_BLOB;
  if (blobsBundle.proofs.length !== expectedProofsLength) {
    throw Error(
      `Invalid proofs length for BlobsBundleV2 format: expected ${expectedProofsLength}, got ${blobsBundle.proofs.length}`
    );
  }

  const commitmentBytes = blobsBundle.commitments.flatMap((commitment) => Array(CELLS_PER_EXT_BLOB).fill(commitment));
  const cellIndices = Array.from({length: blobsBundle.blobs.length}).flatMap(() =>
    Array.from({length: CELLS_PER_EXT_BLOB}, (_, i) => i)
  );
  const cells = blobsBundle.blobs.flatMap((blob) => ckzg.computeCells(blob));
  const proofBytes = blobsBundle.proofs.flat();

  try {
    ckzg.verifyCellKzgProofBatch(commitmentBytes, cellIndices, cells, proofBytes);
  } catch {
    throw new Error("Error in verifyCellKzgProofBatch");
  }
```

hughy replied:
Updated to check pre-Fulu first and return early 👍

matthewkeil commented on packages/beacon-node/src/execution/engine/http.ts:
```diff
@@ -415,14 +415,23 @@ export class ExecutionEngineHttp implements IExecutionEngine {
+        break;
+      default:
+        method = "engine_getPayloadV1";
+    }
```
Am thinking we might want to swap around the condition so that future forks dont require maintenance of this method.  What do you think?

```suggestion
    switch (fork) {
      case ForkName.phase0:
      case ForkName.altair:
      case ForkName.bellatrix:
        method = "engine_getPayloadV1";
        break;
      case ForkName.capella:
        method = "engine_getPayloadV2";
        break;
      case ForkName.deneb:
        method = "engine_getPayloadV3";
        break;
      case ForkName.electra:
        method = "engine_getPayloadV4";
        break;
      default:
        method = "engine_getPayloadV5";
        break;
    }
```

hughy committed:
cache computed cells in producedContentsCache

update type of cached values to include optional cells

update input to validateBlobsAndKzgCommitments to include optional cells, recompute cells if not present

try to read cached contents before assembling data column sidecars. block contents in publishBlock come from validator process and will not include the cells, so we need to read these from the cached produced contents

fix lint errors

hughy commented on packages/beacon-node/src/api/impl/beacon/blocks/index.ts:
```diff
@@ -87,7 +87,10 @@ export function getBeaconBlockApi({
       if (isForkPostFulu(fork)) {
-        dataColumnSidecars = computeDataColumnSidecars(config, signedBlock, signedBlockOrContents);
+        const blockHash = toRootHex((signedBlock.message.body as deneb.BeaconBlockBody).executionPayload.blockHash);
+        const cachedContents = chain.producedContentsCache.get(blockHash);
```
@g11tech please correct me if I'm misunderstanding, but I think we need to add this cache read here since `signedBlockOrContents` comes from a validator API call and won't include the cells

g11tech replied:
yes thats correct

hughy committed:
update fulu types for cell proofs

define BlobsBundle based on fulu KZGProofs type

define ExecutionPayloadAndBlobsBundle based on fulu type

g11tech commented on packages/beacon-node/src/api/impl/beacon/blocks/index.ts:
```diff
@@ -87,7 +87,10 @@ export function getBeaconBlockApi({
       let blockData: BlockInputAvailableData;
       if (isForkPostFulu(fork)) {
-        dataColumnSidecars = computeDataColumnSidecars(config, signedBlock, signedBlockOrContents);
+        const blockHash = toRootHex((signedBlock.message.body as deneb.BeaconBlockBody).executionPayload.blockHash);
```
shouldn't you use chain.getContents since you already modified thaat as well

hughy committed:
read from producedContentsCache using getContents

hughy committed:
add unit tests for computeDataColumnSidecars

update unit tests for validateBlobsAndKzgCommitments to cover case where cells
are passed

hughy committed:
update max length of fulu KZGProofs to match spec

use FIELD_ELEMENTS_PER_EXT_BLOB * MAX_BLOB_COMMITMENTS_PER_BLOCK

g11tech commented on packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:
```diff
@@ -347,13 +349,21 @@ export async function produceBlockBody<T extends BlockType>(
+            if (ForkSeq[fork] >= ForkSeq.fulu) {
+              const cells = blobsBundle.blobs.map((blob) => ckzg.computeCells(blob));
+              contents = {...contents, cells};
+              validateBlobsAndKzgCommitments(fork as ForkBlobs, executionPayload, blobsBundle, cells);
```
```suggestion
             if (this.opts.sanityCheckExecutionEngineBlobs){
               validateBlobsAndKzgCommitments(fork as ForkBlobs, executionPayload, blobsBundle, cells);
             }
```

hughy replied:
Updated to remove separate call for `validateBlobsAndKzgCommitments`. Now there is only one call, conditional on `opts.sanityCheckExecutionEngineBlobs`

hughy committed:
make cell proof verification optional

only verify cell proofs during block production if
sanityCheckExecutionEngineBlobs option is set

matthewkeil committed:
fix: reverse order of fork selection switch

hughy committed:
replace usage of ForkBlobs with ForkPostDeneb

hughy committed:
switch condition order in validateBlobsAndKzgCommitments, return early

reduce nesting depth

matthewkeil commented on packages/params/src/forkName.ts:
```diff
@@ -112,10 +112,10 @@ export function isForkPostFulu(fork: ForkName): fork is ForkPostFulu {
 export type ForkPostBlobs = ForkPostFulu;
-export type ForkBlobs = ForkName.deneb | ForkName.electra;
-export const forkBlobs = [ForkName.deneb, ForkName.electra];
+export type ForkBlobs = ForkName.deneb | ForkName.electra | ForkName.fulu;
```
please remove the fulu from all of these here

matthewkeil reviewed (approved):
LGTM!!  🚀 

I'm going to approve but will leave final approval to @g11tech before merging

matthewkeil reviewed (approved):
LGTM!!  🚀 

I'm going to approve but will leave final approval to @g11tech before merging

hughy committed:
revert changes to forkBlobs

hughy committed:
add missing semicolon

hughy committed:
fix lint error in validateBlobsAndKzgCommitments.test.ts

organize imports from '@lodestar/types'

g11tech commented on packages/beacon-node/src/chain/produceBlock/validateBlobsAndKzgCommitments.ts:
```diff
@@ -1,16 +1,54 @@
+        `Invalid proofs length for BlobsBundleV1 format: expected ${blobsBundle.blobs.length}, got ${blobsBundle.proofs.length}`
+      );
+    }
+    return;
```
lets add blob batch proof validation here as well since we added for cells

hughy replied:
Updated with blob batch proof verification

hughy committed:
add verification of blob proofs

verify blob proofs on v1 blobs bundles in validateBlobsAndKzgCommitments

add unit test for verification of blob proofs

matthewkeil committed:
Merge branch 'peerDAS' into peerDAS-cell-proofs

dguenther commented on packages/types/src/fulu/sszTypes.ts:
```diff
@@ -31,13 +32,17 @@ export const Cell = new ByteVectorType(BYTES_PER_FIELD_ELEMENT * FIELD_ELEMENTS_
     column: DataColumn,
     kzgCommitments: denebSsz.BlobKzgCommitments,
-    kzgProofs: denebSsz.KZGProofs,
+    kzgProofs: KZGProofs,
```
DataColumnSidecar should still be `ListCompositeType(denebSsz.KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK)` https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/das-core.md#datacolumnsidecar

```suggestion
    kzgProofs: denebSsz.KZGProofs,
```

hughy committed:
revert incorrect change to DataColumnSidecar ssz

limit to length of sidecar kzg proofs should be one per blob, not one per cell
per blob

matthewkeil reviewed (approved):
LGTM! 🚀

wemeetagain commented:
:tada: This PR is included in [v1.34.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.34.0) :tada:
